Infima Shuffle Shard Loadbalancer

cgetzen opened the pull request:
Signed-off-by: Charlie Getzen <charliegetzenlc@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: upstream: add shuffle-sharding load balancer
Additional Description:
Risk Level: Medium
Testing: TBD
Docs Changes: TBD
Release Notes: TBD
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
https://github.com/envoyproxy/envoy/issues/14663
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

This implements what is closest to option 1 from the google doc referenced in the issue https://github.com/envoyproxy/envoy/issues/14663.

If I get a tentative thumbs-up, I will begin working on:
- tests
- importing infima from https://github.com/cgetzen/cpp-infima. This is a port of AWS' Route53 Infima, and by importing it from a repo that contains the NOTICE I believe we will be more in-line with licensing. I would be happy to transfer ownership of cpp-infima.
- logging and metrics
- docs 

### Description

Please learn more about shuffle sharding from the linked issue.
Using a Lattice to shuffle shard, we can ensure that a request isn't isolated to a single fault-boundary.

`use_zone_as_dimensions`: this allows one to use envoy's zonality as a Lattice dimension.
`dimensions`: A list of strings that serve as Lattice dimensions. Similar to the SubsetLb, this pulls from ENVOY_LB metadata.
`endpoints_per_cell`: The amount of endpoints returned for a given cell. To calculate the number of endpoints available to a given unique request: endpoints_per_cell x cells. The number of cells is the multiplication of number of dimension values for each dimension (eg. "AZ" could have 3 values, "Region" could have 2 values = 6 cells)

mattklein123 commented:
cc @wgallagher could you take a first pass?

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -887,6 +898,9 @@ message Cluster {
   LbSubsetConfig lb_subset_config = 22;
 
+  // Configuration for the ShuffleShard.
+  LbShuffleShardConfig lb_shuffle_shard_config = 53;
```
As I already mentioned previously, we will want to put this into an extension configuration and not add it to the Cluster.

mattklein123 reviewed (commented):
Just a quick drive by until Bill gets a chance to take a look.

/wait

billgallagher-okta commented on bazel/repositories.bzl:
```diff
@@ -218,6 +219,16 @@ def _boringssl_fips():
         patches = ["@envoy//bazel/external:boringssl_fips.patch"],
     )
 
+def _com_github_cgetzen_cpp_infima():
```
will this move to a public repo before commit?

cgetzen replied:
It is currently public here: https://github.com/cgetzen/cpp-infima
I'd like to see ownership transferred to github.com/envoyproxy, but there may be other options I haven't considered.

mattklein123 replied:
I haven't reviewed the overall PR yet since it's still drafting, but just a heads up but we are unlikely to take a dependency on this external repo. I would just inline whatever code you actually need to make the PR function vs. using the library.

cgetzen replied:
Gotcha. This is a good opportunity for me to learn how licensing works.

billgallagher-okta commented on generated_api_shadow/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -886,6 +897,9 @@ message Cluster {
   // Configuration for load balancing subsetting.
   LbSubsetConfig lb_subset_config = 22;
 
+  // Configuration for the ShuffleShard.
```
per matt's comment above, this should also move to extension

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+    dimensions_.push_back(dimension);
+  if (use_zone_as_dimension_)
+    dimensions_.push_back("_envoy_zone");
+  if (!dimensions_.size())
```
nit: ::empty() instead of ::size()

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+  if (!dimensions_.size())
+    dimensions_.push_back("_envoy_unit_coord");
+
+  lattice_ = new Lattice<HostConstSharedPtr>(dimensions_);
```
this naked pointer should be replaced with a smart pointer. Maybe provider a helper function to build the dimensions and then you can constructor initialize lattice_

billgallagher-okta commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -587,6 +589,35 @@ class RandomLoadBalancer : public ZoneAwareLoadBalancerBase {
+  const bool use_zone_as_dimension_;
+  std::vector<std::string> dimensions_;
+  const bool use_dimensions_;
+  Lattice<Upstream::HostConstSharedPtr>* lattice_;
```
remove naked pointer

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+      use_dimensions_(config.dimensions().size()),
+      shuffle_sharder_(ShuffleSharder<HostConstSharedPtr>(12345)) {
+
+  // ENVOY_LOG(debug, "endpoints_per_cell: {}", endpoints_per_cell_);
```
remove

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+      shuffle_sharder_(ShuffleSharder<HostConstSharedPtr>(12345)) {
+
+  // ENVOY_LOG(debug, "endpoints_per_cell: {}", endpoints_per_cell_);
+  // ENVOY_LOG(debug, "use_zone_as_dimension: {}", use_zone_as_dimension_);
```
remove

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+      endpoints_per_cell_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, endpoints_per_cell, 2)),
+      use_zone_as_dimension_(config.use_zone_as_dimension()),
+      use_dimensions_(config.dimensions().size()),
+      shuffle_sharder_(ShuffleSharder<HostConstSharedPtr>(12345)) {
```
remove magic number. Should be a constant?

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+      continue;
+    }
+
+    std::vector<HostConstSharedPtr> hosts;
```
seems odd to create a separate vector for each host?

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+
+  if (use_dimensions_) {
+    if (!host->metadata()) {
+      // ENVOY_LOG(warn, "ignoring host {} because it has no metadata", host->hostname());
```
remove

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+    const auto& filter_it =
+        metadata.filter_metadata().find(Envoy::Config::MetadataFilters::get().ENVOY_LB);
+    if (filter_it == metadata.filter_metadata().end()) {
+      // ENVOY_LOG(warn, "ignoring host {} because it has no envoy.lb metadata", host->hostname());
```
remove

billgallagher-okta commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+  if (use_dimensions_) {
+    if (!host->metadata()) {
+      // ENVOY_LOG(warn, "ignoring host {} because it has no metadata", host->hostname());
+      return {};
```
does this return an empty optional or an optional with an empty list?

cgetzen replied:
My understanding is that this returns an empty optional. It looks like `std::nullopt` is used in other portions of envoy and may be more idiomatic. Will switch.

cgetzen commented:
- [x] Move to extension 🎉 (may need more discussion on layout)
- [x] Removed unnecessary comments around ENVOY_LOG (may need more discussion - using ENVOY_LOG_MISC)
- [x] More idiomatic empty optionals
- [x] Moved magic number into constant (may need more discussion)
- [x] Removed naked pointer

mattklein123 commented on CODEOWNERS:
```diff
@@ -170,5 +170,7 @@ extensions/filters/http/oauth2 @rgs1 @derekargueta @snowp
 /*/extensions/matching/input_matchers/consistent_hashing @snowp @donyu
 # user space socket pair and event
 /*/extensions/io_socket/user_space @lambdai @antoniovicente
+# load balancers
+/*/extensions/load_balancers @snowp @mattklein123
```
I haven't reviewed the change, but if you did go ahead and make an LB extension point to fix that issue, please PR that first and we can review independently. Would be great to get that fixed! (This PR is too large as it is). Thanks!

/wait
